soc: cc13x2_cc26x2: P (Power Amplifier) SubGhz radio support + CC1352P Launchpad board

niflostancu opened the pull request:
Hello,

This PR is a follow-up for #38503 (which was closed due to inactivity).
Updated and tested (using echo client / server) for the latest Zephyr development branch (2022-05-25).

---
Added two new SOC definitions (for the CC13x2/CC26x2 P chip variants) and modified the IEEE802154 Sub-GHz driver to support high-power 20dBm TX (for CC1352P only).
Plus some enhancements for further RF driver configurability (ability to override RF setup + antenna callbacks).

I also created a new board (cc1352p_launchxl) with RF antenna switching implementation (based on RF event callbacks, as done in TI samples).
Tested using two of these boards and the echo client / server samples (as expected, RSSI greatlty improved).

niflostancu committed:
soc: cc13xx_cc26xx: config option to enable boost mode

New KConfig option to set `CCFG_FORCE_VDDR_HH` inside CC13xx/CC26xx
customer configuration file, making it it possible to use 14dBm
TX power (instead of the default 13 dBm limitation).

Signed-off-by: Stancu Florin <niflostancu@gmail.com>

niflostancu committed:
boards: cc1352p1_launchxl: new board, support antenna switching

Added new CC1352P1 LaunchXL board supporting 20dBm TX for its
sub-GHz radio.

Note that the board has a multiplexer circuit to switch between
2.4GHz, High-Power TX and Sub1GHz states, for which a custom board
module was implemented, together with board-specific device-tree
bindings and pinctrl definitions for each of the RF states.

Signed-off-by: Stancu Florin <niflostancu@gmail.com>

niflostancu committed:
soc: cc13xx_cc26xx: add P chip variants to Kconfig

Add SOC_CC1352P and SOC_CC2652P chip types to SoC's Kconfig
(with integrated high power amplifiers).

Also requires modifications to HAL TI's family conditions.

Signed-off-by: Stancu Florin <niflostancu@gmail.com>

niflostancu committed:
soc: cc13xx_cc26xx: add Kconfig option for custom radio hwattrs

Useful for implementing board-specific antenna switching using the RF
event callback.

Signed-off-by: Stancu Florin <niflostancu@gmail.com>

niflostancu committed:
drivers: ieee802154: cc13xx_cc26xx_subg: PA TX amplifier support

Enhance IEEE802154 Sub-GHz driver to support CC13x2P's internal power
amplifier (20dBm) for TX.

Note: requires board-specific antenna switching for it to work
properly.

Signed-off-by: Stancu Florin <niflostancu@gmail.com>

cfriedt commented on drivers/ieee802154/ieee802154_cc13xx_cc26xx_subg.c:
```diff
@@ -68,14 +85,135 @@ static uint32_t overrides_sub_ghz[] = {
+
+/* Sub GHz power tables */
+#if defined(CONFIG_IEEE802154_CC13XX_CC26XX_SUB_GHZ_CUSTOM_POWER_TABLE)
+extern RF_TxPowerTable_Entry iee802154_cc13xx_subg_power_table[];
```
minor typo: `iee802154_cc13xx_subg_power_table` -> `ieee802154_cc13xx_subg_power_table`

cfriedt reviewed (dismissed):
LGTM. Just 1 minor typo that I saw. Thanks for resubmitting 👍 
The `hal_ti` change has been merged, so don't forget to update `west.yml`

gmarull commented on boards/arm/cc1352p1_launchxl/board_antenna.c:
```diff
@@ -0,0 +1,154 @@
+	ARG_UNUSED(dev);
+
+	/* set all paths to low */
+	IOCPinTypeGpioOutput(BOARD_RF_24GHZ);
+	GPIO_setOutputEnableDio(BOARD_RF_24GHZ, GPIO_OUTPUT_DISABLE);
+	IOCPinTypeGpioOutput(BOARD_RF_HIGH_PA);
+	GPIO_setOutputEnableDio(BOARD_RF_HIGH_PA, GPIO_OUTPUT_DISABLE);
+	IOCPinTypeGpioOutput(BOARD_RF_SUB1GHZ);
+	GPIO_setOutputEnableDio(BOARD_RF_SUB1GHZ, GPIO_OUTPUT_DISABLE);
```
- why isn't this using GPIO API? 
- why isn't this described in Devicetree?
- Shouldn't these tasks be done by a driver? Does every board need to copy this whole code?

gmarull commented on boards/arm/cc1352p1_launchxl/board_antenna.c:
```diff
@@ -0,0 +1,154 @@
+			if (paType == RF_TxPowerTable_HighPA) {
+				/* PA enable --> HIGH PA */
+				/* LNA enable --> Sub-1 GHz */
+				IOCPortConfigureSet(BOARD_RF_24GHZ, IOC_PORT_GPIO,
+						IOC_IOMODE_NORMAL);
+				/* Note: RFC_GPO3 is a work-around because the RFC_GPO1 */
+				/* is sometimes not de-asserted on CC1352 Rev A. */
+				IOCPortConfigureSet(BOARD_RF_HIGH_PA,
+						IOC_PORT_RFC_GPO3, IOC_IOMODE_NORMAL);
+				IOCPortConfigureSet(BOARD_RF_SUB1GHZ,
+						IOC_PORT_RFC_GPO0, IOC_IOMODE_NORMAL);
+			} else {
+				/* RF core active --> Sub-1 GHz */
+				IOCPortConfigureSet(BOARD_RF_24GHZ,
+						IOC_PORT_GPIO, IOC_IOMODE_NORMAL);
+				IOCPortConfigureSet(BOARD_RF_HIGH_PA,
+						IOC_PORT_GPIO, IOC_IOMODE_NORMAL);
+				IOCPortConfigureSet(BOARD_RF_SUB1GHZ,
+						IOC_PORT_GPIO, IOC_IOMODE_NORMAL);
+				GPIO_setOutputEnableDio(BOARD_RF_SUB1GHZ, GPIO_OUTPUT_ENABLE);
```
same here

gmarull commented on boards/arm/cc1352p1_launchxl/cc1352p1_launchxl.dts:
```diff
@@ -0,0 +1,110 @@
+/dts-v1/;
+
+#include <ti/cc1352r.dtsi>
+#include "../cc1352r1_launchxl/boosterpack_connector.dtsi"
+#include "../cc1352r1_launchxl/cc1352r1_launchxl-pinctrl.dtsi"
```
each board needs to have its own pinctrl file.

gmarull commented on boards/arm/cc1352p1_launchxl/cc1352p1_launchxl.dts:
```diff
@@ -0,0 +1,110 @@
+#include "../cc1352r1_launchxl/boosterpack_connector.dtsi"
+#include "../cc1352r1_launchxl/cc1352r1_launchxl-pinctrl.dtsi"
+
+#define BTN_GPIO_FLAGS (GPIO_ACTIVE_LOW | GPIO_PULL_UP)
```
we normally don't do this, just add the appropriate flags to the gpio flags cell.

gmarull commented on boards/arm/cc1352p1_launchxl/board_antenna.c:
```diff
@@ -0,0 +1,154 @@
+ * switch.
+ */
+
+#include <zephyr.h>
+#include <device.h>
+#include <init.h>
```
please use <zephyr/...> prefix. why are `<zephyr.h>` and `<device.h>` needed?

niflostancu commented:
Oh f**, misclick :D

niflostancu commented:
Hello again. Now that the new version has been released, I would like to finish with this PR :D 

So.. the only thing left was the usage of TI's low-level HAL API for GPIO / IOC and definition of GPIO pin number constants for some board-specific code (antenna switch based on radio mode).

OFC, the preferred way would have been to make use of Zephyr's API & Device Tree features, so tell me if the following this is okay:
- I'll define several `<board>_pinctrl.dtsi` entries for the three antenna MUX inputs and each of their states (subghz / 2.4GHz / power amplifier TX), which would map to the appropiate pins;
- modify `board_antenna.c` and use Zephyr's the GPIO & PinCtrl (importing the `struct pinctrl_dev_config` from the devicetree statically / by name) API.
- the overall structure of the code (the RF status conditions) will remain the same;

But it would still be a board-specific module. Is this okay?

cfriedt commented:
> Hello again. Now that the new version has been released, I would like to finish with this PR :D
> 
> So.. the only thing left was the usage of TI's low-level HAL API for GPIO / IOC and definition of GPIO pin number constants for some board-specific code (antenna switch based on radio mode).
> 
> OFC, the preferred way would have been to make use of Zephyr's API & Device Tree features, so tell me if the following this is okay:
> 
> * I'll define several `<board>_pinctrl.dtsi` entries for the three antenna MUX inputs and each of their states (subghz / 2.4GHz / power amplifier TX), which would map to the appropiate pins;

I'm thinking it might make sense to create dt-bindings for this purpose as well as a dedicated node in the tree. Suggested bindings:
1. antenna-mux-gpio-ctl (a phandle - exact wording is 100% flexible)
2. antenna-mux-gpio-pin (would refer to a specific pin number inside of the given gpio controller

Not sure if it's possible, but a build-time assertion would be ideal to ensure that the specified pin is valid.

> * modify `board_antenna.c` and use Zephyr's the GPIO & PinCtrl (importing the `struct pinctrl_dev_config` from the devicetree statically / by name) API.
> * the overall structure of the code (the RF status conditions) will remain the same;
> 
> But it would still be a board-specific module. Is this okay?

Definitely would be good to use devicetree.

@gmarull - do you have any additional feedback?

niflostancu commented:
Hello, I managed to reimplement the antenna switching using Zephyr's devicetree & pinctrl features.
I might need to further rename the bindings to match the ones in your PR (I saw it just now).
Let me know if you have any suggestions.
Thanks.

niflostancu commented on boards/arm/cc1352p1_launchxl/board_antenna.c:
```diff
@@ -90,76 +101,35 @@ void board_cc13xx_rf_callback(RF_Handle client, RF_GlobalEvent events, void *arg
-				IOCPortConfigureSet(BOARD_RF_SUB1GHZ,
-						IOC_PORT_GPIO, IOC_IOMODE_NORMAL);
-				GPIO_writeDio(BOARD_RF_SUB1GHZ, 1);
+				pinctrl_apply_state(ant_pcfg, PINCTRL_STATE_ANT_SUBG);
+				/* Manually set the sub-GHZ antenna switch DIO */
+				gpio_pin_configure_dt(&ant_gpios[BOARD_ANT_GPIO_SUBG], 1);
```
I need it to switch the DIO pin back and forth from GPIO to the RF core (`IOC_PORT_RFC_GPO*`) states (see `<board>_pinctrl.dtsi`).

niflostancu committed:
boards: cc1352r1_launchxl: fix echo samples

Recent Zephyr changes (IEEE 802.15.4 config moved from KConfig
to DeviceTree) left the TI CC1352r1_launchxl board's net examples in
a non-working state, mainly due to the fact that CC13x2 has two
IEEE802154 interfaces available (2.4GHz and sub-GHz) and none were
enabled.

This commit enables the 2.4GHz radio inside the board's DTS and also
introduces a new devicetree overlay inside the echo* samples:
'boards/boards/cc1352-enable-subg.overlay', enabling the sub-GHz
interface.

Signed-off-by: Stancu Florin <niflostancu@gmail.com>

niflostancu commented:
Hello, I rebased it and fixed the conflicts, though I haven't touched the pinctrl-based antenna switch code from months ago.

Additionally, I prepended a new commit to fix the cc1352r1 board config and the echo_* samples with sub-GHz overlays from the latest IEEE 802.15.4 Kconfig to DTS migration and re-tested its functionality.

I will attach here some experimental results using the echo_* samples with sub-ghz radio cofigured using various TX parameters:
<details>
  <summary>CC1352P1 LaunchXL experimental logs</summary>

```
Echo Server logs for CC1352P1, IEEE802154g channel 0 (868 MHz):
Positioning: ~2 meters apart, antennas fixed with tape ;)

1. 14dBm configured, without BOOST MODE (radio actually transmits at 13dBm)

<dbg> net_echo_server_sample: process_udp: (udp6): UDP (IPv6): Received and replied with 19 bytes
<dbg> ieee802154_cc13xx_cc26xx_subg: ieee802154_cc13xx_cc26xx_subg_rx_done: Received: len = 114, rssi = -57 status = 0
<dbg> ieee802154_cc13xx_cc26xx_subg: ieee802154_cc13xx_cc26xx_subg_rx_done: Received: len = 44, rssi = -57 status = 0
<dbg> net_echo_server_sample: process_udp: (udp6): UDP (IPv6): Received and replied with 66 bytes
<dbg> ieee802154_cc13xx_cc26xx_subg: ieee802154_cc13xx_cc26xx_subg_rx_done: Received: len = 98, rssi = -56 status = 0
<dbg> net_echo_server_sample: process_udp: (udp6): UDP (IPv6): Received and replied with 36 bytes
<dbg> ieee802154_cc13xx_cc26xx_subg: ieee802154_cc13xx_cc26xx_subg_rx_done: Received: len = 80, rssi = -57 status = 0
=> RSSI = -57

2. 14dBm configured, with BOOST MODE patch

<dbg> ieee802154_cc13xx_cc26xx_subg: ieee802154_cc13xx_cc26xx_subg_rx_done: Received: len = 61, rssi = -50 status = 0
<dbg> net_echo_server_sample: process_udp: (udp6): UDP (IPv6): Received and replied with 83 bytes
<dbg> ieee802154_cc13xx_cc26xx_subg: ieee802154_cc13xx_cc26xx_subg_rx_done: Received: len = 105, rssi = -50 status = 0
<dbg> net_echo_server_sample: process_udp: (udp6): UDP (IPv6): Received and replied with 43 bytes
<dbg> ieee802154_cc13xx_cc26xx_subg: ieee802154_cc13xx_cc26xx_subg_rx_done: Received: len = 114, rssi = -50 status = 0
<dbg> ieee802154_cc13xx_cc26xx_subg: ieee802154_cc13xx_cc26xx_subg_rx_done: Received: len = 58, rssi = -50 status = 0
=> RSSI = -50

3. 20dBm configured, High-Power TX activated

<dbg> net_echo_server_sample: process_udp: (udp6): UDP (IPv6): Received and replied with 19 bytes
<dbg> ieee802154_cc13xx_cc26xx_subg: ieee802154_cc13xx_cc26xx_subg_rx_done: Received: len = 88, rssi = -25 status = 0
<dbg> net_echo_server_sample: process_udp: (udp6): UDP (IPv6): Received and replied with 26 bytes
<dbg> ieee802154_cc13xx_cc26xx_subg: ieee802154_cc13xx_cc26xx_subg_rx_done: Received: len = 114, rssi = -25 status = 0
<dbg> ieee802154_cc13xx_cc26xx_subg: ieee802154_cc13xx_cc26xx_subg_rx_done: Received: len = 78, rssi = -25 status = 0
<dbg> net_echo_server_sample: process_udp: (udp6): UDP (IPv6): Received and replied with 100 bytes
=> RSSI = -25
```
</details>

Any suggestions are welcome ;) thanks.

cfriedt commented on drivers/ieee802154/Kconfig.cc13xx_cc26xx:
```diff
@@ -91,4 +91,19 @@ config IEEE802154_CC13XX_CC26XX_SUB_GHZ_RADIO_TX_RETRIES
+	bool "Use custom power table structure (advanced)"
+	default n
+	help
+	  Imports the radio power table from an external module.
```
We should really put this in Device Tree

niflostancu replied:
My use case here was for an application to configure custom / proprietary RF parameters (e.g., different bit rates / modulations for better signal resiliency, as generated by SmartRF Studio).
However, a Device Tree Binding for this needs to model all of the RF properties (and there are dozens). But I guess it would still be more elegant than our current approach.

niflostancu commented on drivers/ieee802154/ieee802154_cc13xx_cc26xx_subg.c:
```diff
@@ -68,14 +85,135 @@ static uint32_t overrides_sub_ghz[] = {
+extern RF_TxPowerTable_Entry ieee802154_cc13xx_subg_power_table[];
+
+#elif defined(CONFIG_SOC_CC1352R)
+static const RF_TxPowerTable_Entry ieee802154_cc13xx_subg_power_table[] = {
```
Hmm, that's a pretty neat idea, but I'd rather do it in a subsequent pull request.

cfriedt replied:
My thoughts exactly. It's a separate topic

niflostancu commented on drivers/ieee802154/ieee802154_cc13xx_cc26xx_subg.c:
```diff
@@ -68,14 +85,135 @@ static uint32_t overrides_sub_ghz[] = {
+};
+
+/* Radio setup command for CC1312P / CC1352P */
+static volatile rfc_CMD_PROP_RADIO_DIV_SETUP_PA_t ieee802154_cc13xx_subg_radio_div_setup = {
```
My use case here was for an application to configure custom / proprietary RF parameters (e.g., different bit rates / modulations for better signal resiliency, as generated by SmartRF Studio).
However, a Device Tree Binding for this needs to model all of the RF properties (and there are dozens). But I guess it would still be more elegant than our current approach.

carlescufi commented:
@rlubos could you take a look?

niflostancu commented:
Thanks, I will rebase, re-test it using my launchxls and update the PR tomorrow (I don't have them at my current location) ;)

jadonk commented:
@cfriedt @niflostancu @gmarull @rlubos The system seems to say that one of @gmarull change requests weren't addressed, but it seems @niflostancu replied to all change requests, so I'm missing why this is blocked. I was looking to see if there was anything I could help address to assist with this PR, but it seems to all be done. Once this is done, I can look at the CC1352P7 support and then do a new PR for the BeagleConnect Freedom (now CC1352P7-based).

Please let me know what is blocking.

cfriedt commented:
> Please let me know what is blocking.

@jadonk - it's an annoying quirk with GitHub. Reviewers do not get a notification unless explicitly "at-ed" or the "re-request" button is pressed in the `...` menu beside their names.

I just did it on behalf of @niflostancu.

gmarull commented on boards/arm/cc1352p1_launchxl/board_antenna.c:
```diff
@@ -0,0 +1,165 @@
+	IOCPinTypeGpioOutput(BOARD_RF_HIGH_PA);
+	GPIO_setOutputEnableDio(BOARD_RF_HIGH_PA, GPIO_OUTPUT_ENABLE);
+	IOCPinTypeGpioOutput(BOARD_RF_SUB1GHZ);
+	GPIO_setOutputEnableDio(BOARD_RF_SUB1GHZ, GPIO_OUTPUT_ENABLE);
```
please use GPIO API and Devicetree to specify antenna settings. -1 on using HAL in tree for configuring GPIOs.

niflostancu replied:
Hello, I actually did that in my last commit: https://github.com/zephyrproject-rtos/zephyr/pull/46050/commits/ceaf79b53a06a6e90a057ac060f0b765fd47ee43

gmarull replied:
please squash changes

niflostancu replied:
Alright, I will do a rebase a bit later today and re-test on my boards. Thanks.

niflostancu commented:
Ahaaa, found the re-request review ;) lesson learned!

cfriedt reviewed (dismissed):
LGTM - @gmarull PTAL

jadonk commented:
I don't know those acronyms. Getting the needed review?

gmarull commented on boards/arm/cc1352p1_launchxl/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+    board_antenna.c
+    )
+
+zephyr_library_compile_definitions("DeviceFamily_CC13X2")
```
this is not needed, see https://github.com/zephyrproject-rtos/zephyr/pull/48369

gmarull commented on boards/arm/cc1352p1_launchxl/board_antenna.c:
```diff
@@ -0,0 +1,135 @@
+	ARG_UNUSED(dev);
+	int i;
+
+	/* default pinctrl configuration: set all antenna mux control pins as GPIOs */
+	pinctrl_apply_state(ant_pcfg, PINCTRL_STATE_DEFAULT);
```
this all continues to look like a hack to me.

niflostancu replied:
The antenna switch pins must either be asserted as GPIOs or re-mapped to the chip's radio core based on the current RF configuration scenario. Thus, Zephyr's GPIO & PinCtrl APIs must be simultaneously used to re-arrange the antenna pathways. Although it might seem a bit hackish since it is not the intended use case of this API (though it seems okay for me), it was done to avoid using the low-level HAL routines (as requested earlier in this PR).

gmarull commented on boards/arm/cc1352p1_launchxl/board_antenna.c:
```diff
@@ -0,0 +1,135 @@
+		RF_RadioSetup *setupCommand = (RF_RadioSetup *)arg;
+
+		switch (setupCommand->common.commandNo) {
+		case (CMD_RADIO_SETUP):
+		case (CMD_BLE5_RADIO_SETUP):
+			loDivider = RF_LODIVIDER_MASK & setupCommand->common.loDivider;
+			break;
+		case (CMD_PROP_RADIO_DIV_SETUP):
```
please no `()` here

gmarull commented on boards/arm/cc1352p1_launchxl/cc1352p1_launchxl.dts:
```diff
@@ -0,0 +1,147 @@
+	 * 20 dBm TX   0     1     0     // DIO29 mux to IOC_PORT_RFC_GPO3 for auto
+	 */
+	antenna_mux0: antenna_mux0 {
+		compatible = "ti,antenna-mux";
```
antenna switch (SKY13317-373LF), so compatible looks wrong.

niflostancu replied:
True, renamed binding to `skyworks,sky13317` (kept private to `cc1352p` board)

jadonk commented:
Hey @gmarull , it seems your changes were done. Is there a chance you can review and approve?

jadonk commented:
Ping @gmarull

jadonk commented:
@gmarull If changes are still required to the GPIO/PINCTRL usage, can you make some recommendation? It isn't clear to me that you have any alternative in mind.

cfriedt commented:
@rlubos - please ack if you have a moment.

cfriedt commented:
Including the previous iteration of this PR, it took a full year for this change to get in. We really need to do better :-/

niflostancu commented:
Hi, thanks all for helping merge this.

I'm a somewhat active user of this platform (though it may take some time to migrate to the newer Zephyr releases), but feel free to tag me whenever you need real hardware testing.

As a next step, I plan to update the TI SDK to the latest version (I also see they published a new [Matter-based SDK](https://github.com/TexasInstruments/matter/) on GitHub, worth checking out), though I don't have any timeline yet.

cfriedt commented:
> > Including the previous iteration of this PR, it took a full year for this change to get in. We really need to do better :-/
> 
> This PR had some important issues that took a long time to fix (and what was merged isn't ideal either, but at least it is localized on a board). Also, TI is a platform that is kind of unmaintained, one has to check what is the real state of a platform before using it.

I'm hoping @vaishnavachath can step in to maintain TI platforms. Thanks for helping review @gmarull

vaishnavachath commented:
> > > Including the previous iteration of this PR, it took a full year for this change to get in. We really need to do better :-/
> > 
> > 
> > This PR had some important issues that took a long time to fix (and what was merged isn't ideal either, but at least it is localized on a board). Also, TI is a platform that is kind of unmaintained, one has to check what is the real state of a platform before using it.
> 
> I'm hoping @vaishnavachath can step in to maintain TI platforms. Thanks for helping review @gmarull

@cfriedt submitted PR for https://github.com/zephyrproject-rtos/zephyr/pull/54324 for adding myself as collaborator, please review. I have some WIP changes to add mcuboot support for the Simplelink platforms, CC1352X7 (704KB flash, 144 KB SRAM) version, will come back with those in sometime :smile:  .

@niflostancu , Thank you for the contributions, I have tested the changes on a CC1352P1 launchpad and also a custom board with CC1352P, PA changes work great.
